From 1f564154bf5c723e24489d28c845275fc01c774c Mon Sep 17 00:00:00 2001
From: spl <simon@raspberrypi.org>
Date: Fri, 13 Nov 2015 16:11:52 +0000
Subject: [PATCH 2/2] Non thread-safe GTK+ calls removed to main thread.

---
 src/rc_gui.c | 20 +++++++++++++-------
 1 file changed, 13 insertions(+), 7 deletions(-)

--- a/src/rc_gui.c
+++ b/src/rc_gui.c
@@ -79,7 +79,7 @@
 static GObject *loclang_cb, *loccount_cb, *locchar_cb;
 static GObject *language_ls, *country_ls;
 
-static GtkWidget *main_dlg;
+static GtkWidget *main_dlg, *msg_dlg;
 
 /* Initial values */
 
@@ -385,6 +385,12 @@
 	g_signal_connect (loccount_cb, "changed", G_CALLBACK (country_changed), NULL);
 }
 
+static gboolean close_msg (gpointer data)
+{
+	gtk_widget_destroy (GTK_WIDGET (msg_dlg));
+	return FALSE;
+}
+
 static gpointer locale_thread (gpointer data)
 {
     char buffer[256];
@@ -392,7 +398,7 @@
     system ("sudo locale-gen");
     sprintf (buffer, "sudo update-locale LANG=%s", glocale);
     system (buffer);
-	gtk_widget_destroy (GTK_WIDGET (data));
+    g_idle_add (close_msg, NULL);
     return NULL;
 }
 
@@ -555,7 +561,7 @@
                 }
 
                 // warn about a short delay...
-                GtkWidget *msg_dlg = (GtkWidget *) gtk_dialog_new ();
+                msg_dlg = (GtkWidget *) gtk_dialog_new ();
                 gtk_window_set_title (GTK_WINDOW (msg_dlg), "");
                 gtk_window_set_modal (GTK_WINDOW (msg_dlg), TRUE);
                 gtk_window_set_decorated (GTK_WINDOW (msg_dlg), FALSE);
@@ -570,7 +576,7 @@
 	            gtk_widget_show_all (msg_dlg);
 
                 // launch a thread with the system call to update the generated locales
-                g_thread_new (NULL, locale_thread, msg_dlg);
+                g_thread_new (NULL, locale_thread, NULL);
 
                 // set reboot flag
                 needs_reboot = 1;
@@ -640,7 +646,7 @@
 static gpointer timezone_thread (gpointer data)
 {
     system ("sudo dpkg-reconfigure --frontend noninteractive tzdata");
-	gtk_widget_destroy (GTK_WIDGET (data));
+    g_idle_add (close_msg, NULL);
     return NULL;
 }
 
@@ -715,7 +721,7 @@
             system (buffer);
 
             // warn about a short delay...
-            GtkWidget *msg_dlg = (GtkWidget *) gtk_dialog_new ();
+            msg_dlg = (GtkWidget *) gtk_dialog_new ();
             gtk_window_set_title (GTK_WINDOW (msg_dlg), "");
             gtk_window_set_modal (GTK_WINDOW (msg_dlg), TRUE);
             gtk_window_set_decorated (GTK_WINDOW (msg_dlg), FALSE);
@@ -730,7 +736,7 @@
             gtk_widget_show_all (msg_dlg);
 
             // launch a thread with the system call to update the timezone
-            g_thread_new (NULL, timezone_thread, msg_dlg);
+            g_thread_new (NULL, timezone_thread, NULL);
         }
 	}
 	gtk_widget_destroy (dlg);
